node/libs/protobuf_build/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+// Imports accessed from the generated code.
+pub use prost;
+use prost::Message as _;
+pub use prost_reflect;
+use std::{fs, path::Path, sync::Mutex};
+pub use syntax::*;
+
+mod canonical;
+mod ident;
+mod syntax;
+
+/// Traversed all the files in a directory recursively.
+fn traverse_files(
+    path: &Path,
+    f: &mut impl FnMut(&Path) -> anyhow::Result<()>,
+) -> anyhow::Result<()> {
+    if !path.is_dir() {
+        f(path).with_context(|| path.to_str().unwrap().to_string())?;
+        return Ok(());
+    }
+    for entry in fs::read_dir(path)? {
+        traverse_files(&entry?.path(), f)?;
+    }
+    Ok(())
+}
+
+/// Protobuf descriptor + info about the mapping to rust code.
+pub struct Descriptor {
+    /// Root proto package that all proto files in this descriptor belong to.
+    /// Rust types have been generated relative to this root.
+    proto_root: ProtoName,
+    /// Raw descriptor proto.
+    descriptor_proto: prost_types::FileDescriptorSet,
+    /// Direct dependencies of this descriptor.
+    dependencies: Vec<&'static Descriptor>,
+}
+
+impl Descriptor {
+    /// Constructs a Descriptor.
+    pub fn new(
+        proto_root: ProtoName,
+        dependencies: Vec<&'static Descriptor>,
+        descriptor_bytes: &impl AsRef<[u8]>,
+    ) -> Self {
+        Descriptor {
+            proto_root,
+            dependencies,
+            descriptor_proto: prost_types::FileDescriptorSet::decode(descriptor_bytes.as_ref())
+                .unwrap(),
+        }
+    }
+
+    /// Loads the descriptor to the pool, if not already loaded.
+    pub fn load(&self, pool: &mut prost_reflect::DescriptorPool) -> anyhow::Result<()> {
+        if self
+            .descriptor_proto
+            .file
+            .iter()
+            .all(|f| pool.get_file_by_name(f.name()).is_some())
+        {
+            return Ok(());
+        }
+        for d in &self.dependencies {
+            d.load(pool)?;
+        }
+        pool.add_file_descriptor_set(self.descriptor_proto.clone())?;
+        Ok(())
+    }
+
+    /// Loads the descriptor to the global pool and returns a copy of the global pool.
+    pub fn load_global(&self) -> prost_reflect::DescriptorPool {
+        /// Global descriptor pool.
+        static POOL: Lazy<Mutex<prost_reflect::DescriptorPool>> = Lazy::new(Mutex::default);
+        let pool = &mut POOL.lock().unwrap();
+        self.load(pool).unwrap();
+        pool.clone()
+    }
+}
+
+/// Code generation config. Use it in build scripts.
+pub struct Config {
+    /// Input directory relative to $CARGO_MANIFEST_DIR with the proto files to be compiled.
+    pub input_root: InputPath,
+    /// Implicit prefix that should be prepended to proto paths of the proto files in the input directory.
+    pub proto_root: ProtoPath,
+    /// Descriptors of the dependencies and the rust absolute paths under which they will be available from the generated code.
+    pub dependencies: Vec<(RustName, &'static Descriptor)>,
```

Comment:
Maybe, a stupid ask: Theoretically, we can add crate name (via `env!("CARGO_CRATE_NAME")`) and module path (via [`module_path!`](https://doc.rust-lang.org/std/macro.module_path.html)) information to `Descriptor` instances, so that each descriptor knows where it's located. In this case, `RustName` could be avoided. (There are caveats, such as renaming the package dependency [using `package = ...`](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#renaming-dependencies-in-cargotoml).)

Reply:
I tried that and it works, but making the path to the proto location public as-is looks very inconvenient to me - more likely the generated proto will be exported via "pub use" statement somewhere. Also, as you say, we need a rename functionality anyway, so making things explicit looks like a better option to me.